File: core/src/main/java/com/ustadmobile/core/controller/WebChunkPresenter.java
package com.ustadmobile.core.controller;

import com.ustadmobile.core.db.UmAppDatabase;
import com.ustadmobile.core.db.dao.ContentEntryDao;
import com.ustadmobile.core.impl.UmAccountManager;
import com.ustadmobile.core.impl.UmCallback;
import com.ustadmobile.core.impl.UstadMobileSystemImpl;
import com.ustadmobile.core.util.ContentEntryUtil;
import com.ustadmobile.core.util.UMFileUtil;
import com.ustadmobile.core.view.ContentEntryDetailView;
import com.ustadmobile.core.view.ContentEntryListView;
import com.ustadmobile.core.view.DummyView;
import com.ustadmobile.core.view.WebChunkView;
import com.ustadmobile.lib.db.entities.ContentEntry;

import java.util.Hashtable;

import static com.ustadmobile.core.impl.UstadMobileSystemImpl.ARG_REFERRER;
import static com.ustadmobile.core.view.WebChunkView.ARG_CHUNK_PATH;
import static com.ustadmobile.core.view.WebChunkView.ARG_CONTENT_ENTRY_ID;

public class WebChunkPresenter extends UstadBaseController<WebChunkView> {

    private final WebChunkView viewContract;
    private String navigation;

    public WebChunkPresenter(Object context, Hashtable arguments, WebChunkView view) {
        super(context, arguments, view);
        this.viewContract = view;
    }

    @Override
    public void onCreate(Hashtable savedState) {
        super.onCreate(savedState);
        UmAppDatabase repoAppDatabase = UmAccountManager.getRepositoryForActiveAccount(getContext());
        ContentEntryDao contentEntryDao = repoAppDatabase.getContentEntryDao();

        Long entryUuid = Long.valueOf((String) getArguments().get(ARG_CONTENT_ENTRY_ID));
        navigation = (String) getArguments().get(ARG_REFERRER);

        contentEntryDao.getContentByUuid(entryUuid, new UmCallback<ContentEntry>() {
            @Override
            public void onSuccess(ContentEntry result) {
                viewContract.setToolbarTitle(result.getTitle());
            }

            @Override
            public void onFailure(Throwable exception) {

            }
        });

        view.mountChunk((String) getArguments().get(ARG_CHUNK_PATH), new UmCallback<String>() {
            @Override
            public void onSuccess(String firstUrl) {
                view.loadUrl(firstUrl);
            }

            @Override
            public void onFailure(Throwable exception) {

            }
        });
    }

    public void handleUrlLinkToContentEntry(String sourceUrl) {
        UstadMobileSystemImpl impl = UstadMobileSystemImpl.getInstance();
        UmAppDatabase repoAppDatabase = UmAccountManager.getRepositoryForActiveAccount(getContext());

        ContentEntryUtil.goToContentEntryByViewDestination(
                sourceUrl,
                repoAppDatabase, impl,
                true,
                getContext(), new UmCallback<Void>() {
                    @Override
                    public void onSuccess(Void result) {

                    }

                    @Override
                    public void onFailure(Throwable exception) {
                        viewContract.showError(exception.getMessage());
                    }
                });
    }

    public void handleUpNavigation() {
        UstadMobileSystemImpl impl = UstadMobileSystemImpl.getInstance();
        String lastEntryListArgs = UMFileUtil.getLastReferrerArgsByViewname(ContentEntryDetailView.VIEW_NAME, navigation);
        if (lastEntryListArgs != null) {
            impl.go(ContentEntryDetailView.VIEW_NAME,
                    UMFileUtil.parseURLQueryString(lastEntryListArgs), view.getContext(),
                    UstadMobileSystemImpl.GO_FLAG_CLEAR_TOP | UstadMobileSystemImpl.GO_FLAG_SINGLE_TOP);
        } else {
            impl.go(DummyView.VIEW_NAME,
                    null, view.getContext(),
                    UstadMobileSystemImpl.GO_FLAG_CLEAR_TOP | UstadMobileSystemImpl.GO_FLAG_SINGLE_TOP);
        }

    }

}

File: core/src/main/java/com/ustadmobile/core/controller/WebChunkPresenter.kt
package com.ustadmobile.core.controller

import com.ustadmobile.core.generated.locale.MessageID
import com.ustadmobile.core.impl.NoAppFoundException
import com.ustadmobile.core.impl.UmAccountManager
import com.ustadmobile.core.impl.UmCallback
import com.ustadmobile.core.impl.UstadMobileSystemImpl
import com.ustadmobile.core.impl.UstadMobileSystemImpl.Companion.ARG_REFERRER
import com.ustadmobile.core.util.ContentEntryUtil
import com.ustadmobile.core.util.UMFileUtil
import com.ustadmobile.core.view.ContentEntryDetailView
import com.ustadmobile.core.view.DummyView
import com.ustadmobile.core.view.WebChunkView
import com.ustadmobile.core.view.WebChunkView.Companion.ARG_CONTAINER_UID
import com.ustadmobile.core.view.WebChunkView.Companion.ARG_CONTENT_ENTRY_ID
import com.ustadmobile.lib.db.entities.Container
import com.ustadmobile.lib.db.entities.ContentEntry

class WebChunkPresenter(context: Any, arguments: Map<String, String>, view: WebChunkView)
    : UstadBaseController<WebChunkView>(context, arguments, view) {

    private var navigation: String? = null

    override fun onCreate(savedState: Map<String, String?>?) {
        super.onCreate(savedState)
        val repoAppDatabase = UmAccountManager.getRepositoryForActiveAccount(context)
        val contentEntryDao = repoAppDatabase.contentEntryDao
        val containerDao = repoAppDatabase.containerDao

        val entryUuid = arguments.getValue(ARG_CONTENT_ENTRY_ID)!!.toLong()
        val containerUid = arguments.getValue(ARG_CONTAINER_UID)!!.toLong()


        navigation = arguments[ARG_REFERRER] ?: ""

        contentEntryDao.getContentByUuid(entryUuid, object : UmCallback<ContentEntry> {
            override fun onSuccess(result: ContentEntry?) {
                view.runOnUiThread(Runnable { view.setToolbarTitle(result!!.title) })
            }

            override fun onFailure(exception: Throwable?) {
                view.runOnUiThread(Runnable { view.showError(UstadMobileSystemImpl.instance.getString(MessageID.error_opening_file, context)) })
            }
        })

        containerDao.findByUid(containerUid, object : UmCallback<Container> {

            override fun onSuccess(result: Container?) {
                view.mountChunk(result!!, object : UmCallback<String> {
                    override fun onSuccess(result: String?) {
                        if (result != null) {
                            view.loadUrl(result)
                        } else {
                            view.runOnUiThread(Runnable { view.showError(UstadMobileSystemImpl.instance.getString(MessageID.error_opening_file, context)) })
                        }
                    }

                    override fun onFailure(exception: Throwable?) {
                        view.runOnUiThread(Runnable { view.showError(UstadMobileSystemImpl.instance.getString(MessageID.error_opening_file, context)) })
                    }
                })
            }

            override fun onFailure(exception: Throwable?) {

            }
        })
    }

    fun handleUrlLinkToContentEntry(sourceUrl: String) {
        val impl = UstadMobileSystemImpl.instance
        val repoAppDatabase = UmAccountManager.getRepositoryForActiveAccount(context)

        ContentEntryUtil.goToContentEntryByViewDestination(
                sourceUrl,
                repoAppDatabase, impl,
                true,
                context, object : UmCallback<Any> {
            override fun onSuccess(result: Any?) {

            }

            override fun onFailure(exception: Throwable?) {
                if (exception != null) {
                    val message = exception.message
                    if (exception is NoAppFoundException) {
                        view.runOnUiThread(Runnable {
                            view.showErrorWithAction(impl.getString(MessageID.no_app_found, context),
                                    MessageID.get_app,
                                    exception.mimeType!!)
                        })
                    } else {
                        view.runOnUiThread(Runnable { view.showError(message!!) })
                    }
                }
            }
        })
    }

    fun handleUpNavigation() {
        val impl = UstadMobileSystemImpl.instance
        val lastEntryListArgs = UMFileUtil.getLastReferrerArgsByViewname(ContentEntryDetailView.VIEW_NAME, navigation!!)
        if (lastEntryListArgs !=
                null) {
            impl.go(ContentEntryDetailView.VIEW_NAME,
                    UMFileUtil.parseURLQueryString(lastEntryListArgs), view.context,
                    UstadMobileSystemImpl.GO_FLAG_CLEAR_TOP or UstadMobileSystemImpl.GO_FLAG_SINGLE_TOP)
        } else {
            impl.go(DummyView.VIEW_NAME, mapOf(), view.context,
                    UstadMobileSystemImpl.GO_FLAG_CLEAR_TOP or UstadMobileSystemImpl.GO_FLAG_SINGLE_TOP)
        }

    }

}
